src/dmd/argtypes_sysv_x64.d

```diff
@@ -0,0 +1,394 @@
+        return new TypeTuple();
+
+    const classification = classify(t, size);
+    const classes = classification.slice();
+    const c0 = classes[0];
+
+    if (c0 == Class.Memory)
+         return new TypeTuple();
+    if (c0 == Class.X87)
+        return new TypeTuple(Type.tfloat80);
+    if (c0 == Class.ComplexX87)
+        return new TypeTuple(Type.tfloat80, Type.tfloat80);
+
+    if (classes.length > 2 ||
+        (classes.length == 2 && classes[1] == Class.SSEUp))
+    {
+        assert(c0 == Class.SSE);
+        foreach (c; classes[1 .. $])
+            assert(c == Class.SSEUp);
+
+        assert(size % 8 == 0);
+        return new TypeTuple(new TypeVector(Type.tfloat64.sarrayOf(classes.length)));
+    }
+
+    assert(classes.length >= 1 && classes.length <= 2);
+    Type[2] argtypes;
+    foreach (i, c; classes)
+    {
+        // the last eightbyte may be filled partially only
+        auto sizeInEightbyte = (i < classes.length - 1) ? 8 : size % 8;
+        if (sizeInEightbyte == 0)
+            sizeInEightbyte = 8;
+
+        if (c == Class.Integer)
+        {
+            argtypes[i] =
+                sizeInEightbyte > 4 ? Type.tint64 :
+                sizeInEightbyte > 2 ? Type.tint32 :
+                sizeInEightbyte > 1 ? Type.tint16 :
+                                      Type.tint8;
+        }
+        else if (c == Class.SSE)
+        {
+            argtypes[i] =
+                sizeInEightbyte > 4 ? Type.tfloat64 :
+                                      Type.tfloat32;
+        }
+        else
+            assert(0, "Unexpected class");
+    }
+
+    return classes.length == 1
+        ? new TypeTuple(argtypes[0])
+        : new TypeTuple(argtypes[0], argtypes[1]);
+}
+
+
+private:
+
+// classification per eightbyte (64-bit chunk)
+enum Class : ubyte
+{
+    Integer,
+    SSE,
+    SSEUp,
+    X87,
+    X87Up,
+    ComplexX87,
+    NoClass,
+    Memory
+}
+
+Class merge(Class a, Class b)
+{
+    if (a == b)
+        return a;
+    if (a == Class.NoClass)
+        return b;
+    if (b == Class.NoClass)
+        return a;
+    if (a == Class.Memory || b == Class.Memory)
+        return Class.Memory;
+    if (a == Class.Integer || b == Class.Integer)
+        return Class.Integer;
+    if (a == Class.X87 || b == Class.X87 ||
+        a == Class.X87Up || b == Class.X87Up ||
+        a == Class.ComplexX87 || b == Class.ComplexX87)
+        return Class.Memory;
+    return Class.SSE;
+}
+
+struct Classification
+{
+    Class[4] classes;
+    int numEightbytes;
+
+    const(Class[]) slice() const { return classes[0 .. numEightbytes]; }
+}
+
+Classification classify(Type t, size_t size)
+{
+    scope v = new ToClassesVisitor(size);
+    t.accept(v);
+    return Classification(v.result, v.numEightbytes);
+}
+
+extern (C++) final class ToClassesVisitor : Visitor
+{
+    const size_t size;
+    const int numEightbytes;
+    Class[4] result = Class.NoClass;
+
+    this(size_t size)
+    {
+        assert(size > 0);
```

Comment:
Perhaps this should be in a contract instead, i.e. an `in` block.

Reply:
I don't think contract are used in dmd.

Reply:
Is there a reason we can't start 😃.